File: o11nplugin-contrail-vro-custom/src/main/java/net/juniper/contrail/vro/CustomManagedType.java
/*
 * Copyright (c) 2018 Juniper Networks, Inc. All rights reserved.
 */

package net.juniper.contrail.vro;

import com.vmware.o11n.sdk.modeldriven.extension.Interceptor;
import com.vmware.o11n.sdk.modeldrivengen.model.*;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class CustomManagedType extends ManagedType {
    private final ManagedType delegate;

    public CustomManagedType(ManagedType type) {
        this.delegate = type;

        generateRefsMethods();
    }

    public static CustomManagedType wrap(ManagedType type) {
        return new CustomManagedType(type);
    }

    public List<CustomRefsField> getRefsFields() {
        List<Field> fields= Arrays.asList(delegate.getModelClass().getDeclaredFields());
        return fields.stream()
                .filter( field -> field.getName().endsWith("_refs") )
                .map(CustomRefsField::wrapField)
                .collect(Collectors.toList());
    }

    @Override
    public boolean isGenerate() {
        return delegate.isGenerate();
    }

    @Override
    public void setGenerate(boolean generate) {
        delegate.setGenerate(generate);
    }

    @Override
    public boolean isFindable() {
        return delegate.isFindable();
    }

    @Override
    public void setFindable(boolean findable) {
        delegate.setFindable(findable);
    }

    @Override
    public boolean isRootIdPropagated() {
        return delegate.isRootIdPropagated();
    }

    @Override
    public void setRootIdPropagated(boolean rootIdPropagated) {
        delegate.setRootIdPropagated(rootIdPropagated);
    }

    @Override
    public String getFullSuperClassName() {
        return delegate.getFullSuperClassName();
    }

    @Override
    public void setFullSuperClassName(String fullSuperClassName) {
        delegate.setFullSuperClassName(fullSuperClassName);
    }

    @Override
    public String getPackageName() {
        return delegate.getPackageName();
    }

    @Override
    public String getClassName() {
        return delegate.getClassName();
    }

    @Override
    public String getName() {
        return delegate.getName();
    }

    @Override
    public boolean isSingleton() {
        return delegate.isSingleton();
    }

    @Override
    public void setSingleton(boolean singeton) {
        delegate.setSingleton(singeton);
    }

    @Override
    public boolean isCreateable() {
        return delegate.isCreateable();
    }

    @Override
    public void setCreateable(boolean createable) {
        delegate.setCreateable(createable);
    }

    @Override
    public String getFullClassName() {
        return delegate.getFullClassName();
    }

    @Override
    public void setFullClassName(String className) {
        delegate.setFullClassName(className);
    }

    @Override
    public List<ManagedMethod> getMethods() {
        return delegate.getMethods();
    }

    @Override
    public void setMethods(List<ManagedMethod> methods) {
        delegate.setMethods(methods);
    }

    @Override
    public String getDoc() {
        return delegate.getDoc();
    }

    @Override
    public void setDoc(String doc) {
        delegate.setDoc(doc);
    }

    @Override
    public List<ManagedConstructor> getConstructors() {
        return delegate.getConstructors();
    }

    @Override
    public void setConstructors(List<ManagedConstructor> constructors) {
        delegate.setConstructors(constructors);
    }

    @Override
    public String toString() {
        return delegate.toString();
    }

    @Override
    public void setName(String name) {
        delegate.setName(name);
    }

    @Override
    public Class<?> getModelClass() {
        return delegate.getModelClass();
    }

    @Override
    public void setModelClass(Class<?> wrappedClass) {
        delegate.setModelClass(wrappedClass);
    }

    @Override
    public Class<? extends Interceptor> getInterceptor() {
        return delegate.getInterceptor();
    }

    @Override
    public void setInterceptor(Class<? extends Interceptor> interceptorClass) {
        delegate.setInterceptor(interceptorClass);
    }

    private void generateRefsMethods() {
        List methods = this.getMethods();

        for (CustomRefsField customField : this.getRefsFields()) {
            String name = "get" + customField.getMethodName();
            String propertyName = Character.toLowerCase(customField.getMethodName().charAt(0)) + customField.getMethodName().substring(1);
            FormalParameter returnParameter = createReturnsFormalParameter();

            ManagedMethod method = new ManagedMethod();
            method.setName(name, name);
            method.setOriginalPropertyName(propertyName);
            method.setPropertyName(propertyName);
            method.setParams(new ArrayList<>());
            method.setIsInheritedWrapperMethod(true);
            method.setPropertyReadOnly(true);
            method.setReturns(returnParameter);

            methods.add(method);
        }
    }

    private FormalParameter createReturnsFormalParameter() {
        FormalParameter returnParameter = new FormalParameter();
        returnParameter.setName("_result");
        returnParameter.setModelType(String.class);
        returnParameter.setFullClassName(String.class.getName());
        returnParameter.setTypeName("String");
        returnParameter.setWrapped(false);
        return returnParameter;
    }
}

File: plugin/o11nplugin-contrail-custom/src/main/kotlin/net/juniper/contrail/vro/CustomManagedType.kt
/*
 * Copyright (c) 2018 Juniper Networks, Inc. All rights reserved.
 */

package net.juniper.contrail.vro

import com.vmware.o11n.sdk.modeldriven.extension.Interceptor
import com.vmware.o11n.sdk.modeldrivengen.model.FormalParameter
import com.vmware.o11n.sdk.modeldrivengen.model.ManagedConstructor
import com.vmware.o11n.sdk.modeldrivengen.model.ManagedMethod
import com.vmware.o11n.sdk.modeldrivengen.model.ManagedType
import java.util.ArrayList

class CustomManagedType(private val delegate: ManagedType) : ManagedType() {

    val refsFields: List<CustomRefsField>
        get() {
            val fields = delegate.modelClass.declaredFields
            return fields
                .asSequence()
                .filter { it.name.endsWith("back_refs") }
                .map { CustomRefsField.wrapField(it) }
                .toList()
        }

    init {
        generateRefsMethods()
    }

    override fun isGenerate(): Boolean =
        delegate.isGenerate

    override fun setGenerate(generate: Boolean) {
        delegate.isGenerate = generate
    }

    override fun isFindable(): Boolean
        = delegate.isFindable

    override fun setFindable(findable: Boolean) {
        delegate.isFindable = findable
    }

    override fun isRootIdPropagated(): Boolean =
        delegate.isRootIdPropagated


    override fun setRootIdPropagated(rootIdPropagated: Boolean) {
        delegate.isRootIdPropagated = rootIdPropagated
    }

    override fun getFullSuperClassName(): String? =
        delegate.fullSuperClassName


    override fun setFullSuperClassName(fullSuperClassName: String) {
        delegate.fullSuperClassName = fullSuperClassName
    }

    override fun getPackageName(): String? =
        delegate.packageName

    override fun getClassName(): String? =
        delegate.className

    override fun getName(): String? =
        delegate.name

    override fun isSingleton(): Boolean =
        delegate.isSingleton

    override fun setSingleton(singeton: Boolean) {
        delegate.isSingleton = singeton
    }

    override fun isCreateable(): Boolean =
        delegate.isCreateable

    override fun setCreateable(createable: Boolean) {
        delegate.isCreateable = createable
    }

    override fun getFullClassName(): String? =
        delegate.fullClassName

    override fun setFullClassName(className: String) {
        delegate.fullClassName = className
    }

    override fun getMethods(): MutableList<ManagedMethod?> =
        delegate.methods

    override fun setMethods(methods: List<ManagedMethod>) {
        delegate.methods = methods
    }

    override fun getDoc(): String? =
        delegate.doc

    override fun setDoc(doc: String) {
        delegate.doc = doc
    }

    override fun getConstructors(): List<ManagedConstructor?>? =
        delegate.constructors

    override fun setConstructors(constructors: List<ManagedConstructor>) {
        delegate.constructors = constructors
    }

    override fun toString(): String =
        delegate.toString()


    override fun setName(name: String) {
        delegate.name = name
    }

    override fun getModelClass(): Class<*>? =
        delegate.modelClass


    override fun setModelClass(wrappedClass: Class<*>) {
        delegate.modelClass = wrappedClass
    }

    override fun getInterceptor(): Class<out Interceptor>? =
        delegate.interceptor


    override fun setInterceptor(interceptorClass: Class<out Interceptor>) {
        delegate.interceptor = interceptorClass
    }

    private fun generateRefsMethods() {
        val methods = this.methods

        for (customField in this.refsFields) {
            val name = "get" + customField.wrapperMethodName
            val propertyName = customField.wrapperMethodName.decapitalize()
            val returnParameter = createReturnsFormalParameter()

            val method = ManagedMethod()
            method.setName(name, name)
            method.originalPropertyName = propertyName
            method.propertyName = propertyName
            method.params = ArrayList()
            method.setIsInheritedWrapperMethod(true)
            method.isPropertyReadOnly = true
            method.returns = returnParameter

            methods.add(method)
        }
    }

    private fun createReturnsFormalParameter(): FormalParameter {
        val returnParameter = FormalParameter()
        returnParameter.name = "_result"
        returnParameter.modelType = String::class.java
        returnParameter.fullClassName = String::class.java.name
        returnParameter.typeName = "String"
        returnParameter.isWrapped = false
        return returnParameter
    }

    companion object {

        fun wrap(type: ManagedType): CustomManagedType {
            return CustomManagedType(type)
        }
    }
}
